Update FAQ.md wrt performance estimates and strategies (#3246)
Chocobozzz/PeerTube: FAQ.md

## Should I have a big server to run PeerTube?

PeerTube should run happily on a vm with 2 threads, at least 1 Gb of RAM and enough storage for videos. In terms of network, a lot will depend on which PeerTube instances you federate with.

As a real life example, the PeerTube demonstration server [https://peertube.cpy.re](https://peertube.cpy.re) runs on 2 vCores and 2GB of RAM. Average consumption is:
 * **CPU**: nginx ~ 20%, peertube ~ 10%,   postgres ~ 1%, redis ~ 3%
 * **RAM**: nginx ~ 6MB, peertube ~ 120MB, postgres ~ 10MB, redis ~ 5MB

> ### CPU

Reviewer: ```suggestion
### CPU

```

Except for video transcoding, a PeerTube instance is not really cpu bound. Neither nginx, peertube, postgres nor redis require a lot of cpu. If it were only for those, once could easily get by with just one thread.

> You will hugely benefit from at least a second thread though, because of transcoding. Transcoding _is_ very cpu intensive. It serves two purposes on a PeerTube instance: it ensures all videos can be played optimally in the web interface, and it generates different resolutions for the same video. PeerTube support for offloading transcoding to other machines is planned, but not yet implemented. See https://github.com/Chocobozzz/PeerTube/issues/1192 .

Author: Thank you for your hint. I checked #938, but that one is about gpu based (hardware accellerated) transcoding. The relevant thing when dimensioning is offloading to different machines. That is an entirely different issue...

> You will hugely benefit from at least a second thread though, because of transcoding. Transcoding _is_ very cpu intensive. It serves two purposes on a PeerTube instance: it ensures all videos can be played optimally in the web interface, and it generates different resolutions for the same video. PeerTube support for offloading transcoding to other machines is planned, but not yet implemented. See https://github.com/Chocobozzz/PeerTube/issues/1192 .

Reviewer 2: Ah, sorry, I misread :)

> You will hugely benefit from at least a second thread though, because of transcoding. Transcoding _is_ very cpu intensive. It serves two purposes on a PeerTube instance: it ensures all videos can be played optimally in the web interface, and it generates different resolutions for the same video. PeerTube support for offloading transcoding to other machines is planned, but not yet implemented. See https://github.com/Chocobozzz/PeerTube/issues/1192 .

Reviewer 3: @markvdb I'm sorry if I misunderstand you, but #1192 is just about transcode prioritization and not offloading?

> You will hugely benefit from at least a second thread though, because of transcoding. Transcoding _is_ very cpu intensive. It serves two purposes on a PeerTube instance: it ensures all videos can be played optimally in the web interface, and it generates different resolutions for the same video. PeerTube support for offloading transcoding to other machines is planned, but not yet implemented. See https://github.com/Chocobozzz/PeerTube/issues/1192 .

Reviewer 3: #947 is discussing offloading

> You will hugely benefit from at least a second thread though, because of transcoding. Transcoding _is_ very cpu intensive. It serves two purposes on a PeerTube instance: it ensures all videos can be played optimally in the web interface, and it generates different resolutions for the same video. PeerTube support for offloading transcoding to other machines is planned, but not yet implemented. See https://github.com/Chocobozzz/PeerTube/issues/1192 .

Reviewer: @markvdb @kontrollanten to be clear, because an issue mentions the problem and contributors debate it doesn't mean it is on the project's roadmap. This issue in particular isn't. It doesn't mean it isn't important, just that we cannot make promises as to when/if we can allocate ressources to tackle its implementation.

> ### RAM

Reviewer: ```suggestion
### RAM

```

1 Gb of RAM should be plenty for a basic PeerTube instance. The only reason you'd want significantly more would be to RAM cache very popular video fragments.

> ### Storage

Reviewer: ```suggestion
### Storage

```

There are two important angles to storage: disk space usage and sustained read speed.

To make a rough estimate of your disk space usage requirements, you want to know the answer to three questions:
- What is the size of the video source archive?
- Do you want to enable transcoding? If so, do you want to provide multiple resolutions per video? Try this out with a few videos and you'll get an idea of a multiplication factor.
- Which sharing mechanisms do you want to enable? Just WebTorrent, or also HLS with p2p? If you want both, this will double your storage needs.

In terms of read speed, you want to make sure that you can saturate your network uplink serving PeerTube videos. This should not be a problem with ssd disks. With traditional hard disks though it is really worth checking! Typical sustained read rates for a well tuned system with a 7200rpm hard disk should hover around 120 MB/s or 960 Mbit/s. This should about saturate a 1 Gbit/s network uplink.

> ### Network

Reviewer: ```suggestion
### Network

```

Reviewer 4: should the storage section include about using s3 compatible solutions as they can help bypass the speed constraints?

Reviewer: @test2a yes

Reviewer: @test2a done

Reviewer 4: @rigelk super.

A rough estimate of a traditional server's video streaming network capacity is usually quite straightforward. You simply divide your server's available bandwidth by the average bandwidth per stream, and you have an upper bound.

> Take a server for example with a 1 Gbps/s uplink for example pushing out 1080p60 streams at 5 Mbps per stream. That means the absolute theoretical upper capacity bound is 200 simultaneous viewers if your server's disk i/o can keep up. Expect a bit less in practice.

Reviewer 2: `1 Gbps` or `1 Gbit/s` - I suggest the use of the `Gbps` form.

Author: Just updated the pull request to reflect your suggestion. Rereading, turns out I misread it and changed everything to consistently use Mbit/s and Gbit/s. That helps keep it consistent with how sustained read rates from disk are noted though...

Reviewer 2: Thank you!

## Should I have a big server to run PeerTube?

> PeerTube should run happily on a vm with 2 threads, at least 1 Gb of RAM and enough storage for videos. In terms of network, a lot will depend on which PeerTube instances you federate with.

Reviewer: ```suggestion
PeerTube should run happily on a virtual machine with 2 threads/vCPUs, at least 1 Gb of RAM and enough storage for videos. In terms of bandwidth, a lot will depend on which PeerTube instances you federate with and what your relation with them is (more about that below).
```

As a real life example, the PeerTube demonstration server [https://peertube.cpy.re](https://peertube.cpy.re) runs on 2 vCores and 2GB of RAM. Average consumption is:
 * **CPU**: nginx ~ 20%, peertube ~ 10%,   postgres ~ 1%, redis ~ 3%
 * **RAM**: nginx ~ 6MB, peertube ~ 120MB, postgres ~ 10MB, redis ~ 5MB

### CPU

> Except for video transcoding, a PeerTube instance is not really cpu bound. Neither nginx, peertube, postgres nor redis require a lot of cpu. If it were only for those, once could easily get by with just one thread.

Reviewer: ```suggestion
Except for video transcoding, a PeerTube instance is not CPU bound. Neither Nginx, PeerTube itself, PpostgreSQL nor Redis require a lot of computing power. If it were only for those, one could easily get by with just one thread/vCPU.
```

You will hugely benefit from at least a second thread though, because of transcoding. Transcoding _is_ very cpu intensive. It serves two purposes on a PeerTube instance: it ensures all videos can be played optimally in the web interface, and it generates different resolutions for the same video. PeerTube support for offloading transcoding to other machines is being discussed, but not yet implemented. See https://github.com/Chocobozzz/PeerTube/issues/947 .

### RAM

> 1 Gb of RAM should be plenty for a basic PeerTube instance. The only reason you'd want significantly more would be to RAM cache very popular video fragments.

Reviewer: ```suggestion
1 Gb of RAM should be plenty for a basic PeerTube instance, which usually takes at most 150 Mb in RAM. The only reason you might want more would be if you colocate your Redis or PostgreSQL services on a non-SSD system.
```
